fft/src/smallft.rs

```diff
@@ -86,109 +89,77 @@ last mod: $Id: smallft.c,v 1.19 2003/10/08 05:12:37 jm Exp $
-            if nr != 0 as c_int {
+            let nq = nl / ntry;
+            let nr = nl - ntry * nq;
+            if nr != 0 {
                 break;
             }
+
             nf += 1;
-            *ifac.offset((nf + 1 as c_int) as isize) = ntry;
-            nl = nq;
-            if !(ntry != 2 as c_int) {
-                if !(nf == 1 as c_int) {
-                    i = 1 as c_int;
-                    while i < nf {
-                        ib = nf - i + 1 as c_int;
-                        *ifac.offset((ib + 1 as c_int) as isize) = *ifac.offset(ib as isize);
-                        i += 1
-                    }
-                    *ifac.offset(2 as c_int as isize) = 2 as c_int
-                }
+            ifac[nf + 1] = ntry;
+            if ntry == 2 && nf != 1 {
+                ifac[1..nf]
+                    .rchunks_exact_mut(2)
+                    .for_each(|chunk| chunk[1] = chunk[0]);
+                ifac[2] = 2;
             }
-            if !(nl != 1 as c_int) {
-                break 'c_10244;
+
+            nl = nq;
+            if nl == 1 {
+                break 'out;
             }
         }
     }
-    *ifac.offset(0 as c_int as isize) = n;
-    *ifac.offset(1 as c_int as isize) = nf;
-    argh = tpi / n as c_float;
-    is = 0 as c_int;
-    nfm1 = nf - 1 as c_int;
-    l1 = 1 as c_int;
-    if nfm1 == 0 as c_int {
+    let nf = nf;
+
+    ifac[0] = n;
+    ifac[1] = nf.try_into().unwrap();
+    if nf == 0 {
         return;
     }
-    k1 = 0 as c_int;
-    while k1 < nfm1 {
-        ip = *ifac.offset((k1 + 2 as c_int) as isize);
-        ld = 0 as c_int;
-        l2 = l1 * ip;
-        ido = n / l2;
-        ipm = ip - 1 as c_int;
-        j = 0 as c_int;
-        while j < ipm {
-            ld += l1;
-            i = is;
-            argld = ld as c_float * argh;
-            fi = 0.0f32;
-            ii = 2 as c_int;
-            while ii < ido {
-                fi += 1.0f32;
-                arg = fi * argld;
-                let fresh0 = i;
-                i = i + 1;
-                *wa.offset(fresh0 as isize) = f64::cos(arg as c_double) as c_float;
-                let fresh1 = i;
-                i = i + 1;
-                *wa.offset(fresh1 as isize) = f64::sin(arg as c_double) as c_float;
-                ii += 2 as c_int
-            }
-            is += ido;
-            j += 1
-        }
-        l1 = l2;
-        k1 += 1
-    }
+
+    let argh = TPI / wa.len() as f32;
+    ifac[2..=nf]
+        .iter()
+        .map(|&ip| ip.try_into().unwrap())
+        .fold((0, 1), |(is, l1), ip: usize| {
+            let l2 = l1 * ip;
+            let ido = wa.len() / l2;
+            let ipm = ip - 1;
+
+            wa[is..]
+                .chunks_exact_mut(ido)
+                .zip((l1..).step_by(l1).map(|ld| ld as f32 * argh))
+                .take(ipm)
+                .for_each(|(chunk, argld)| {
+                    chunk[..(ido - 2).try_into().unwrap_or(0)]
```

Comment:
this is a `saturating_sub()` maybe ?

Reply:
I always forget about `saturating_*` fns...